receiver/sapmreceiver/trace_receiver.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/open-telemetry/opentelemetry-collector/observability"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	jaegertranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace/jaeger"
+	splunksapm "github.com/signalfx/sapm-proto/gen"
+	"github.com/signalfx/sapm-proto/sapmhandler"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+)
+
+const traceSource string = "sapm"
+
+// Receiver type is used to receive spans in the Splunk SAPM format over HTTP
+type sapmReceiver struct {
+	// mu protects the fields of this type
+	mu        sync.Mutex
+	startOnce sync.Once
+	stopOnce  sync.Once
+	logger    *zap.Logger
+
+	config     *Config
+	defaultCtx context.Context
+	server     *http.Server
+
+	nextConsumer consumer.TraceConsumer
+}
+
+// HTTPHandlerFunction returns an http.HandlerFunc that handles SAPM requests
+func (sr *sapmReceiver) HTTPHandlerFunc() func(rw http.ResponseWriter, req *http.Request) {
+	// receiverFn is a discrete function passed to sapm-proto's http.HanlderFunc.  This receiverFn actually processes the
+	// sapm and passes it on to the next trace consumer
+	receiverFn := func(ctx context.Context, sapm *splunksapm.PostSpansRequest, err error) error {
+		rCtx := observability.ContextWithReceiverName(ctx, "sapm")
+
+		// process the sapm
+		if sapm != nil {
+			for _, batch := range sapm.Batches {
+				// convert the jager batches to OCProto
+				td, err := jaegertranslator.ProtoBatchToOCProto(*batch)
+				if err != nil {
+					return err
+				}
+				td.SourceFormat = "sapm"
+
+				// pass the trace data to the next consumer
+				sr.nextConsumer.ConsumeTraceData(rCtx, td)
+
+				// We MUST unconditionally record metrics from this reception.
+				observability.RecordMetricsForTraceReceiver(rCtx, len(batch.Spans), len(batch.Spans)-len(td.Spans))
+			}
+		}
+
+		return nil
+	}
+
+	// get a handler function from the sapm-proto repo and pass it the receiverFn
+	sapmHandlerFn := sapmhandler.NewTraceHandlerV2(receiverFn)
+
+	// return an http.HandlerFunc that traces the request and passes it through the sapm-proto handler
+	return func(rw http.ResponseWriter, req *http.Request) {
+
+		// trace this request
+		_, span := trace.StartSpan(sr.defaultCtx, traceSource)
+		sapmHandlerFn(rw, req)
+		span.End()
+
+		return
+	}
+}
+
+// TraceSource implements receiver.TraceReceiver.TraceSource() and returns a tag describing the source format
+func (sr *sapmReceiver) TraceSource() string {
+	return traceSource
+}
+
+// StartTraceReception starts the sapmReceiver's server
+func (sr *sapmReceiver) StartTraceReception(host receiver.Host) error {
+	sr.mu.Lock()
+	defer sr.mu.Unlock()
+
+	var err = oterr.ErrAlreadyStarted
+	sr.startOnce.Do(func() {
+		var ln net.Listener
+
+		// set up the listener
+		ln, err = net.Listen("tcp", sr.config.Endpoint)
+		if err != nil {
+			err = fmt.Errorf("failed to bind to address %s: %v", sr.config.Endpoint, err)
```

Comment:
Shouldn't we `return` here? Is there a point in continuing and trying to setup the http.Server?

Reply:
added a return